docs/Custom Data.md

```diff
@@ -0,0 +1,185 @@
+# Writing a custom data pipeline
+
+There's a `BaseSpeechDataset` class that can be used as base to load the data.
+The library expects that each element in the dataset will be a tuple (audio_tensor, text_label),
+where the audio tensor has shape (channels, time) and text_label is the corresponding label as a string.
+
+The `BaseSpeechDataset` has two important properties:
+1. A list (or iterable) `.items`, that has all the metadata to load every item in the dataset
+2. The `.loader` module. That is a pytorch class that uses torchaudio to load audio tensors and can apply resampling and mono conversion.
+It was designed to be exported independently of the dataset, so that the same data loading can be used during inference.
+
+To get each element in the dataset, the following code is used, and each function call can be overwritten to control functionality:
+
+```python
+class BaseSpeechDataset(Dataset):
+    def __getitem__(self, index: int) -> Tuple[Tensor, str]:
+        item = self.get_item(index)
+        # Dealing with input
+        audio, sr = self.open_audio(item)
+        audio = self.preprocess_audio(audio, sr)
+        # Dealing with output
+        text = self.open_text(item)
+        text = self.preprocess_text(text)
+        return audio, text
+```
+
+The flow of loading the data happens as follows:
+
+1. `self.get_item` is called with a specific index. It uses `self.items` to return the specific metadata to that example
+2. All the metadata is sent to `self.open_audio`. The relevant subset is used to load the audio tensor and corresponding sample rate, using `self.loader.open_audio(...)`
+3. Inside `self.preprocess_audio` the audio tensor is resampled and converted to mono if necessary using `self.loader.preprocess_audio(...)`.
+At this point, any augmentation that happens at the signal level to individual items can be applied.
+Only the audio tensor is returned, because it's assumed that every audio in the dataset will be resampled to the same sample rate
+4. `self.open_text` uses the same metadata to open the corresponding text label
+5. `self.preprocess_text` can be used to apply any transform directly to the text. Common options are lower case, expanding contractions (`I'm` becomes `I am`), expanding numbers (`42` becomes `forty two`) and removing punctuation
+
+## Example: Loading data from nemo
+
+This example will implement `thunder.data.datamodule.ManifestDatamodule` and `thunder.data.dataset.ManifestSpeechDataset`.
+
+### Load source
+
+The nemo manifest file contains one json in each line, with the relevant data:
```

Comment:
Because it's a bit confusing how NeMo chose to do this, I might elaborate a bit and say something like:

Although NeMo manifest files have a .json extension, they are not actually legal json. They instead follow the [JSON Lines](https://jsonlines.org/) standard where each line is a complete JSON object representing a single audio and its metadata.